### PR TITLE
Running 'ensure_network provider' as a separate job

### PR DESCRIPTION
**Pull request by AlonaKaplan**
'ensure_network' has an api call to ovirt, therefore, it should be run
as a separate jobs to avoid calling it directly from the ui.

All the tests that had to be fixed because of this api call were reverted.

### CODE REVIEW

File: app/models/manageiq/providers/redhat/infra_manager.rb [hunk 1]
```diff
@@ -24,7 +24,21 @@ class ManageIQ::Providers::Redhat::InfraManager < ManageIQ::Providers::InfraMana
 
   before_save :ensure_managers
 
+  def ensure_network_manager_queue()
```
Comment (agrare) [inline app/models/manageiq/providers/redhat/infra_manager.rb]: Don't need the empty parens

File: app/models/manageiq/providers/redhat/infra_manager.rb [hunk 2]
```diff
@@ -24,7 +24,21 @@ class ManageIQ::Providers::Redhat::InfraManager < ManageIQ::Providers::InfraMana
...
+      :class_name  => self.class.name,
+      :instance_id => id,
+      :method_name => 'create_or_update_network_manager',
+      :zone        => my_zone,
+      :args        => []
+    )
+  end
+
   def ensure_network_manager
+    ensure_network_manager_queue()
```
Comment (agrare) [inline app/models/manageiq/providers/redhat/infra_manager.rb]: Same, no parens needed

File: app/models/manageiq/providers/redhat/infra_manager.rb [hunk 3]
```diff
@@ -24,7 +24,21 @@ class ManageIQ::Providers::Redhat::InfraManager < ManageIQ::Providers::InfraMana
 
   before_save :ensure_managers
 
+  def ensure_network_manager_queue()
+    MiqQueue.put(
+      :class_name  => self.class.name,
+      :instance_id => id,
+      :method_name => 'create_or_update_network_manager',
+      :zone        => my_zone,
+      :args        => []
```
Comment (agrare) [inline app/models/manageiq/providers/redhat/infra_manager.rb]: No need to pass empty args, it defaults to an empty array

File: app/models/manageiq/providers/redhat/infra_manager.rb [hunk 4]
```diff
@@ -48,6 +62,7 @@ def ensure_network_manager
 
       if network_manager
         populate_network_manager_connectivity(auth_url)
+        network_manager.save
```
Comment (agrare) [inline app/models/manageiq/providers/redhat/infra_manager.rb]: Can you use `.save!` so this will raise an exception and be logged if saving the network manager fails?

**Comment by pazinio** [comment]
Hi @agarare. This pull request doesn't work well in all the cases. It works well in case the provider already exists. But if I'm in the create flow it doesn't work.
I get - MIQ(MiqQueue#deliver) Message id: [8806], Error: [undefined method `create_or_update_network_manager' for #<Class:0x00000009bb18a0>]
It seems that the infra_manger object is not fully initialized yet. 
It ends up with the workers of the network manager not started until full refresh of the infra provider. 
Do you have an idea how it can be solved?

**Comment by AlonaKaplan** [comment]
Hi @agarare. This pull request doesn't work well in all the cases. It works well in case the provider already exists. But if I'm in the create flow it doesn't work.
I get - MIQ(MiqQueue#deliver) Message id: [8806], Error: [undefined method `create_or_update_network_manager' for #Class:0x00000009bb18a0]
It seems that the infra_manger object is not fully initialized yet.
It ends up with the workers of the network manager not started until full refresh of the infra provider.
Do you have an idea how it can be solved

File: spec/models/manageiq/providers/redhat/infra_manager/refresh/refresher_target_vm_4_spec.rb [hunk 5]
```diff
@@ -3,16 +3,11 @@
 
   before(:each) do
     _, _, zone = EvmSpecHelper.create_guid_miq_server_zone
-    @ems = FactoryGirl.build(:ems_redhat_with_ensure_managers, :zone => zone, :hostname => "192.168.1.105", :ipaddress => "192.168.1.105",
+    @ems = FactoryGirl.create(:ems_redhat, :zone => zone, :hostname => "192.168.1.105", :ipaddress => "192.168.1.105",
```
Comment (borod108) [inline spec/models/manageiq/providers/redhat/infra_manager/refresh/refresher_target_vm_4_spec.rb]: Why is this change required?

**Comment by agrare** [comment]
> I get - MIQ(MiqQueue#deliver) Message id: [8806], Error: [undefined method `create_or_update_network_manager' for #Class:0x00000009bb18a0]
It seems that the infra_manger object is not fully initialized yet.
It ends up with the workers of the network manager not started until full refresh of the infra provider.
Do you have an idea how it can be solved

I got the same thing and its pretty obvious why once you see it haha
`[----] I, [2017-10-23T09:06:47.519370 #16361:2b090b533efc]  INFO -- : MIQ(MiqQueue.put) Message id: [36],  id: [], Zone: [default], Role: [], Server: [], Ident: [generic], Target id: [], Instance id: [], Timeout: [600], Priority: [100], State: [ready], Deliver On: [], Data: [], Args: []`

It is called from before_save, there is no `#id` yet

**Comment by agrare** [comment]
I was talking to @borod108 on gitter and what if we did this, instead of doing this in the before_save/after_save callback, what if we did the check during refresh?  That way if a customer added a network endpoint after they created their InfraManager they don't have to delete&&create the whole provider to pick it up?

**Comment by AlonaKaplan** [comment]
Yes, that is the approach I'm trying now.

**Comment by AlonaKaplan** [comment]
@agrare @Ladas @masayag @borod108 please review

File: app/models/manageiq/providers/redhat/infra_manager.rb [hunk 6]
```diff
@@ -1,3 +1,5 @@
+require 'thread'
```
Comment (agrare) [inline app/models/manageiq/providers/redhat/infra_manager.rb]: What did you need from `thread`?

**Comment by AlonaKaplan** [comment]
No need, leftover. Will remove.

File: spec/models/manageiq/providers/redhat/infra_manager/refresh/refresher_4_1_spec.rb [hunk 7]
```diff
@@ -1,14 +1,13 @@
 describe ManageIQ::Providers::Redhat::InfraManager::Refresh::Refresher do
   before(:each) do
     guid, server, zone = EvmSpecHelper.create_guid_miq_server_zone
-    @ems = FactoryGirl.build(:ems_redhat_with_ensure_managers, :zone => zone, :hostname => "localhost", :ipaddress => "localhost",
+    @ems = FactoryGirl.create(:ems_redhat, :zone => zone, :hostname => "localhost", :ipaddress => "localhost",
```
Comment (borod108) [inline spec/models/manageiq/providers/redhat/infra_manager/refresh/refresher_4_1_spec.rb]: is this change still required?

File: spec/models/manageiq/providers/redhat/infra_manager/refresh/refresher_4_async_spec.rb [hunk 8]
```diff
@@ -1,15 +1,14 @@
 describe ManageIQ::Providers::Redhat::InfraManager::Refresh::Refresher do
   before(:each) do
     _guid, _server, zone = EvmSpecHelper.create_guid_miq_server_zone
-    @ems = FactoryGirl.build(:ems_redhat_with_ensure_managers, :zone => zone, :hostname => "localhost", :ipaddress => "localhost",
+    @ems = FactoryGirl.create(:ems_redhat, :zone => zone, :hostname => "localhost", :ipaddress => "localhost",
```
Comment (borod108) [inline spec/models/manageiq/providers/redhat/infra_manager/refresh/refresher_4_async_spec.rb]: same question

Comment (AlonaKaplan) [reply spec/models/manageiq/providers/redhat/infra_manager/refresh/refresher_4_async_spec.rb]: Previously I changed it to 'build' since I had to mock the 'ovirt_service' result before the 'save', this is not needed any more, so I reverted it back to create.

**Comment by AlonaKaplan** [comment]
depends on https://github.com/ManageIQ/manageiq/pull/16239
